### DISCUSSION TITLE
Timespan literals - like C++ chrono literals

### DISCUSSION THREAD

**Discussion by juwens** [dotnet/csharplang#5076] [Language Ideas]
# Introduction

It would be handy to have const Timespan literals.
In some scenarios (like web/db requests) you will need to specify a quite a few (default) timeout, like 30 seconds, or 3000ms, when calling a method or method parameters

# State of the union

## Nullable Timespan-object.

Pro: 
- explicit Timespan is better than implicit int-milliseconds 
- Timespan Type is more fault resistant for calling with the wrong magnitude (eg 30 as int could be 30 sec or 30 ms or 30 days, etc.)
Con: 
- verbose because Timespan.From... cannot be used as default parameter value
- default value NOT visible with intellisense (mouse pointer hovering over function)
```
void DoStuff(Timespan? timeout = null)
{
    timeout ??= Timespan.FromMilliseconds(300);
}
```

## pure int/long without "hungarian"-time-magnitude-suffix

you see that relatively often, with a hint in the comment, what magnitude the author means, for example Milliseconds

Pro: 
- short
- default value visible with intellisense
- can use default value.
Con:
- an int/long timespan is always kinda ambiguous, can be anything magnitude, ns, µs, ms, s, h, d, etc
```
void DoStuff(int timeout = 300)
{
}
```

## Or to be more explicit:
```
void DoStuff(int timeoutMs = 300)
{
}
```

# Possible literals which are not ambiguous with number literals:

- `5nsec`
- `5µsec
- `5msec``
- `5msec`
- `3min`
- `7hr` or `7hour`

# Bonus: timespan arithmetic with literals should be a relatively simple bonus

```
var timespanSum = 1hr + 12sec + 12msec;
```

# Example literals

```
// simple variable assignment
var timeSpanA = 123msec;

// const variable assignment
const Timespan timeSpanB = 456msec;

// calling a function with Timespan parameter
DoStuff(39sec);

// default method param
void DoStuff(Timespan timeout = 300msec)
{
}
```

# Summary

Advantages:

- can be used in "const" Context, like default method parameter value
- shorter (hence better signal to noise ratio) than TimeSpan.FromMillisecond(123)
- kinda natural language

Disadvantages:

- adds complexity to the C# language
- potentially ambiguous to other literal-suffixes for numbers like single 's', decimal 'm',

**Comment by jnm2** [comment]
In terms of natural language and brevity, what do you think about libraries like <https://github.com/Humanizr/Humanizer#fluent-date> with `2.Minutes()` and `300.Milliseconds()`? (Not to mention `5.Kilobytes()` and so on.)

**Reply by juwens** [reply]
Nice idea for regular code, it wouldn't help for the default parameter and const scenario though.

**Comment by ZacharyPatten** [comment]
I'm not saying this is better than literals built into the language, but this is already valid syntax in C#:
```cs
Time<double> a = (1, Seconds);
Time<double> b = (1, Hours);
Console.WriteLine(a + b); // 3601 Seconds

// you could have abbreviations too
var ms = Milliseconds;
Time<double> c = (1, ms);

// other measurement types
Length<double> length = (1, Meters);
Time<double> time = (1, Seconds);
Speed<double> speed = length / time;
Console.WriteLine(speed); // 1 Meters/Seconds
```

> Source code for examples _**(warning... it is a self-plug/promotion of a work-in-progress concept)**_: https://github.com/ZacharyPatten/Towel/blob/ef0f13506532ed004cce66ab2cca0431d6484df5/Examples/Measurements/Program.cs#L124

Is it kinda hacky? Yes. Can they be defined as `const` values? Not currently no. And obviously `Time<double>` is not `System.TimeSpan`, but you could have an implicit casting operator if desired. Also, the `INumber` interfacing (a preview feature in .NET 6) may prove useful for measurement syntax possibilities in C# in the future.

If the goal of this "timespan literals" discussion is only to reduce syntax, in my opinion there are existing features in the C# language that can get the syntax down to acceptable levels. Thus, the goal should not only be to minimize time-based syntax, but also to provide readable and compile-time safe measurement mathematics for all measurement types (not just for time measurements). If measurement syntax (time literals) were added for time, but not length, mass, speed, etc. I would be disappointed personally. Yet at the same time, even if it was added for other types too, it would likely still not satisfy everyone as there are just so many variables to consider:

- How would operators not available in C# look such as "cross product" or "power"?
- What all measurement types are supported?
- Do you only support metric or US too?
- How much of the metric pre/post-fixes do you support?
- Knots?
- Leagues?
- Cups?
- Gallons?
- Are unit conversions implicit or explicit?
- Are `yottaseconds` supported (if yes... what type? cause it ain't gonna be `TimeSpan`)?
- If conversions are implicit, what unit types are prefered over others? Does it always prefer the smaller unit type in attempts to prevent integer mathematics errors?
- What will the abbreviations be? Is "m/s" "miles per second" or "meters per second"?
- _...list goes on_

Should measurements literals/syntax just be left out of the language altogether and let third party libraries handle it? `¯\_(ツ)_/¯`

Of course lets not forget that F# also has units of measurement syntax. If possible, I would prefer something along those lines compared to only adding `TimeSpan` literals to C# too: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/units-of-measure

**Reply by jnm2** [reply]
> And obviously Time<double> is not System.TimeSpan, but you could have an implicit casting operator if desired.

Wouldn't this operator have to be defined by the System.TimeSpan struct, since it should not be applicable to all possible `Time<T>` instances?

**Reply by ZacharyPatten** [reply]
No, you can make casting operators in generic types just fine. The following is totally valid.
```cs
public struct Time<T>
{
	// ...

	public static implicit operator Time<T>(System.TimeSpan timeSpan)
	{
		return new Time<T>(Convert<double, T>(timeSpan.TotalMilliseconds), Time.Units.Milliseconds);
	}
	
	// ...
}
```

However one thing you **_cannot do but you should be able to do_** is make "type specialized casting operators". You should be able to do the following, but currently is restricted in C#:
```cs
public struct Time<T>
{
	// ...

	public static implicit operator Time<double>(System.TimeSpan timeSpan)
	{
		return new Time<double>(timeSpan.TotalMilliseconds, Time.Units.Milliseconds);
	}
	
	// ...
}
```

In my opinion `CS0556` should be relaxed so that the above code is valid.

**Reply by jnm2** [reply]
I know you can define the operator on `Time<T>`, but I was saying that would likely not be desirable. It is false advertising at compile time and can fail at runtime.

**Reply by ZacharyPatten** [reply]
Well.. yeah code that isn't safe at compile time is bad... that is why we need `INumber` and other features to hurry up. :P

...and relax `CS0556`

**Reply by julealgon** [reply]
I don't quite agree with this part:
> Thus, the goal should not only be to minimize time-based syntax, but also to provide readable and compile-time safe measurement mathematics for all measurement types (not just for time measurements).

While I'm obviously for having the possibility to define units of measurement for other types, conflating everything in the same issue will just cause it to never be tackled.

I think the ability to instantiate `TimeSpan` instances with a constant literal is self-contained enough that one shouldn't necessarily need to add all the other stuff alongside it: they could be made incrementally.

The special UTF8 string literal was added recently and I would consider this somewhat similar in concept.

**Reply by ZacharyPatten** [reply]
> conflating everything in the same issue will just cause it to never be tackled.

I don't think this is true. The C# language design team has some smart cookies on it. :) Once C# measurement syntax is considered a priority topic I don't think it would take an unbearable amount of time for it to be designed and implemented. It just hasn't been a priority topic yet. C# has already been around for 20+ years without it, so I'm willing to wait until it can be done right.

> I think the ability to instantiate TimeSpan instances with a constant literal is self-contained enough that one shouldn't necessarily need to add all the other stuff alongside it: they could be made incrementally.

I disagree that this topic could be implemented "incrementally" without a design that could handle all the measurement types. What does `"m"` mean? Does it mean `"meters"`, `"miles"`, `"milliseconds"`, etc? If we add the `"m"` abbreviation for `"milliseconds"` while implementing timespan syntax it will inevitably conflict with the other measurement types. The `TimeSpan` can't even support all the time measurements. There is no `TimeSpan.FromPicoseconds(...)`. So if we added timespan syntax now without the support for picoseconds or without a design of how to add it in the future, I assume it could not just be done "incrementally" without breaking changes.

This is specifically why I posted my first comment https://github.com/dotnet/csharplang/discussions/5076#discussioncomment-1197445. I wanted to express that I would rather the language team attempt to design measurement syntax that will work for all measurement types rather than just add timespan syntax.

**Reply by julealgon** [reply]
> If we add the `"m"` abbreviation for `"milliseconds"` ...

You'd be abbreviating it wrong, as milliseconds should be `ms` as per the international system of units.

> ...it will inevitably conflict with the other measurement types.

Which you could disambiguate using either target types or casting. We already have ambiguity today with number literals and that's how it works there.

If you were to introduce the capability of specifying "10 milliseconds" as `10ms` today, it doesn't conflict with anything that currently exists. Once a new thing comes up that _would_ conflict, then the conflict can be easily solved at that point.

> The `TimeSpan` can't even support all the time measurements. There is no `TimeSpan.FromPicoseconds(...)`

Again, I don't see this as a problem. They are 2 distinct things. You can implement either one first and it will work out. There is no need to make _everything_ at the same time.

For instance, we could either:
1. allow all existing granularities as literals, then implement `FromPicoseconds` at a later point and add the literal for that as well
2. implement `FromPicoseconds` first as its own standalone feature, and then implement all the literal support including picoseconds

The fact that there is no way to represent picosecods today shouldn't be a blocker for adding time duration literals.

> So if we added timespan syntax now without the support for picoseconds or without a design of how to add it in the future, I assume it could not just be done "incrementally" without breaking changes.

Again, why? What breaking changes are you envisioning? They are again, separate orthogonal concerns, so they can be implemented completely independently of one another.

> I wanted to express that I would rather the language team attempt to design measurement syntax that will work for all measurement types rather than just add timespan syntax.

Combining multiple things into one giant proposal tends to just make it less likely that we'll get _anything_ out of it. Instead, people end up discussing the design endlessly, then discussing "how big/impactful/risky it is" which then just delays the whole implementation several years. I'd rather extract and implement as much as possible in simple valid standalone increments that can more easily be validated and agreed upon in isolation and eventually build up to the "full" support.

**Comment by hez2010** [comment]
Ideally C# should allow all structs to be defined as `const`. See #2878.

**Reply by julealgon** [reply]
Do you think `const structs` would be a prerequisite to allow these literals? We already have things like "decimal literals" today in the language that are actually not constants, but `new Decimal(...)` calls underneath.

While I agree that having `const` structs would be really really nice, I'd say we shouldn't necessarily conflate the ideas here: the timespan literals would still provide a lot of value even if they are not compile-time constants initially.

**Comment by TahirAhmadov** [comment]
The TimeSpan literal idea is such a great idea. This should definitely, definitely be added to the language. 
Even further, I hate all the `int milliseconds` overloads - they should all be `Obsolete`d and only `TimeSpan` overloads should be used. (More of a BCL complaint.)

**Reply by jnm2** [reply]
This is also something you can have today if you write an analyzer. There's also BannedApiAnalyzer if a hardcoded list of APIs is just as good.

**Reply by julealgon** [reply]
@jnm2 such analyzers should only be seen as temporary workarounds. Moving everyone towards `TimeSpan` as the de-facto way of representing "durations of time" should be the ultimate goal and I agree it would be nice to have these `int`-based APIs _actually_ deprecated at some point.
